micrometer-core/src/main/java/io/micrometer/core/instrument/binder/mongodb/MongoMetricsCommandListener.java

```diff
@@ -56,32 +82,123 @@ public MongoMetricsCommandListener(MeterRegistry registry) {
+        }
+        this.collectionNamesCacheMaxSize = builder.collectionNamesCacheMaxSize;
+        this.collectionNamesCacheOverflowErrorTracker = builder.collectionNamesCacheOverflowErrorLogInterval < 1 ? null : new EveryNthOperation(builder.collectionNamesCacheOverflowErrorLogInterval);
     }
 
     @Override
-    public void commandStarted(CommandStartedEvent commandStartedEvent) {
+    public void commandStarted(CommandStartedEvent event) {
+        if (!includeCollectionNameTag) {
+            return;
+        }
+        mongoCommandUtil.determineCollectionName(event.getCommandName(), event.getCommand())
+                .ifPresent(collectionName -> addCollectionNameToCache(event, collectionName));
     }
 
     @Override
     public void commandSucceeded(CommandSucceededEvent event) {
-        timeCommand(event, event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
     @Override
     public void commandFailed(CommandFailedEvent event) {
-        timeCommand(event, event.getElapsedTime(TimeUnit.NANOSECONDS));
+        timeCommand(event, removeCollectionNameFromCache(event), event.getElapsedTime(TimeUnit.NANOSECONDS));
     }
 
-    private void timeCommand(CommandEvent event, long elapsedTimeInNanoseconds) {
-        Timer.builder("mongodb.driver.commands")
-                .description("Timer of mongodb commands")
-                .tags(tagsProvider.commandTags(event))
-                .register(registry)
+    private void timeCommand(CommandEvent event, String collection, long elapsedTimeInNanoseconds) {
+        Timer.Builder timerBuilder = Timer.builder(this.timerMetricName)
+                .description(this.timerMetricDesc)
+                .tags(this.tagsProvider.commandTags(event));
+        if (this.includeCollectionNameTag) {
+            timerBuilder.tag("collection", collection);
+        }
+        timerBuilder
+                .register(this.registry)
                 .record(elapsedTimeInNanoseconds, TimeUnit.NANOSECONDS);
     }
 
+    private void addCollectionNameToCache(CommandEvent event, String collectionName) {
+        if (collectionNamesCache.size() < collectionNamesCacheMaxSize) {
+            collectionNamesCache.put(event.getRequestId(), collectionName);
+            return;
+        }
+        if (collectionNamesCacheOverflowErrorTracker == null) {
+            return;
+        }
+        // Cache over capacity - log every ~Nth for a good balance of warning and not spamming the logs.
+        collectionNamesCacheOverflowErrorTracker.maybeInvoke(() -> LOG.warn("Collection names cache is full - Mongo is not calling listeners properly"));
+    }
+
+    private String removeCollectionNameFromCache(CommandEvent event) {
+        return Optional.ofNullable(collectionNamesCache.remove(event.getRequestId()))
+                .orElse("unknown");
```

Comment:
Usually not using an `Optional` instead of an if block is a good practice, since, here, the Optional is not returned, what do you think about:
```suggestion
        String  collectionName = collectionNamesCache.remove(event.getRequestId());
        return collectionName != null ? collectionName != null : "unknown";
```

Reply:
> Usually not using an Optional instead of an if block is a good practice

I am assuming you meant "is not a good practice" or "is a bad practice".  Personally, I like using Optional but I know we don't use it on Spring Boot in this way and its not a surprise that we would not want to use it here. Adjusting.

Reply:
I guess I overcomplicated it 😄 : I wrote _not_ using an `Optional` is a good practice in these cases (i.e.: doing something else like `if` or `ternary` op.) but, what I simply meant is using an `Optional` is not a good practice in these cases. 🙂

I like `Optional` a lot though in this case I think it is not necessary (it is more useful for return values). There is a talk about [Optional by Stuart Marks](https://www.youtube.com/watch?v=fBYhtvY19xA) where he explains this in greater details.